gobblin-modules/gobblin-eventhub/src/main/java/org/apache/gobblin/eventhub/source/EventhubSource.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.gobblin.source.extractor.extract.AbstractSource;
+import org.apache.gobblin.source.extractor.extract.LongWatermark;
+import org.apache.gobblin.source.extractor.partition.Partition;
+import org.apache.gobblin.source.extractor.partition.Partitioner;
+import org.apache.gobblin.source.workunit.Extract;
+import org.apache.gobblin.source.workunit.MultiWorkUnit;
+import org.apache.gobblin.source.workunit.WorkUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+
+
+/**
+ * Created by Jan on 23-Jan-19.
+ */
+public class EventhubSource extends AbstractSource<Void, EventData> {
+
+    public static final Logger LOG = LoggerFactory.getLogger(EventhubSource.class);
+
+    public static final String WORK_UNIT_STATE_VERSION_KEY = "source.querybased.workUnitState.version";
+    public static final Integer CURRENT_WORK_UNIT_STATE_VERSION = 1;
+
+    @Override
+    public List<WorkUnit> getWorkunits(SourceState state) {
+        List<WorkUnit> workUnits = new ArrayList<>();
+
+        Long previousWatermark = getPreviousWatermark(state);
+        if (previousWatermark == null) {
+            previousWatermark = ConfigurationKeys.DEFAULT_WATERMARK_VALUE;
+        }
+
+        // for now, enforce single work unit to simplify watermark handling:
+        if (state.getPropAsInt(ConfigurationKeys.SOURCE_MAX_NUMBER_OF_PARTITIONS, 1) != 1) {
+            LOG.error("Unsupported number of partitions, resetting to 1.");
+        }
+        state.setProp(ConfigurationKeys.SOURCE_MAX_NUMBER_OF_PARTITIONS, 1);
+        workUnits.addAll(generateWorkUnits(state, previousWatermark));
+
+        LOG.info("Total number of workunits for the current run: " + workUnits.size());
+        List<WorkUnit> previousWorkUnits = this.getPreviousWorkUnitsForRetry(state);
+        LOG.info("Total number of incomplete tasks from the previous run: " + previousWorkUnits.size());
+        workUnits.addAll(previousWorkUnits);
+
+        int numOfMultiWorkunits =
+                state.getPropAsInt(ConfigurationKeys.MR_JOB_MAX_MAPPERS_KEY, ConfigurationKeys.DEFAULT_MR_JOB_MAX_MAPPERS);
+
+        return pack(workUnits, numOfMultiWorkunits);
+    }
+
+    private static List<WorkUnit> pack(List<WorkUnit> workUnits, int numOfMultiWorkunits) {
+        Preconditions.checkArgument(numOfMultiWorkunits > 0);
+
+        if (workUnits.size() <= numOfMultiWorkunits) {
+            return workUnits;
+        }
+        List<WorkUnit> result = Lists.newArrayListWithCapacity(numOfMultiWorkunits);
+        for (int i = 0; i < numOfMultiWorkunits; i++) {
+            result.add(MultiWorkUnit.createEmpty());
+        }
+        for (int i = 0; i < workUnits.size(); i++) {
+            ((MultiWorkUnit) result.get(i % numOfMultiWorkunits)).addWorkUnit(workUnits.get(i));
+        }
+        return result;
+    }
+
+    protected List<WorkUnit> generateWorkUnits(SourceState state, long previousWatermark) {
```

Comment:
I think we should not have a global previousWatermark. If you only support one partition, this will work. But if you have multiple partitions, each partition should have its own previousWatermark. I guess right now your source is much like a QueryBasedSource instead of a partition like source such as KafkaSource. You need to check if you can model it in the same way as KafkaSource, which maintains previousOffsets list for each workunit.